File: bench/go/bits.go
// schema bench — family bits for the Go runner.
//
// Family bits (BENCH-STANDARD.md §1.4): the raw BitWriter/BitReader with the
// 16-width table (227 bits/group) over a 65536-byte buffer. Values vary per
// pass through the LCG (widths are the structure and stay fixed; bytes/pass
// asserted constant); reads rotate 64 pre-written variant buffers, each
// verified to read back exactly what was written before any number is
// produced. Unexported identifiers in the runner's own package per §3.1;
// the timed loops are //go:noinline so the §4.1 verdict has a loop body to
// count.
package main

import (
	"time"

	"github.com/mas-bandwidth/serialize.go"
)

const BitsNumWidths = 16
const BitsBufferSize = 65536

var bitsWidths = [BitsNumWidths]int{1, 32, 7, 13, 3, 25, 8, 19, 4, 28, 11, 16, 2, 30, 6, 22} // 227 bits/group

var gBitsBuffer [BitsBufferSize]byte
var gBitsVariants [NumVariants][BitsBufferSize]byte

func bitsMask(width int) uint32 {
	if width == 32 {
		return 0xFFFFFFFF
	}
	return (1 << width) - 1
}

// the per-pass value variation: one LCG step per pass, values from its bits
func varyBitsValues(values *[BitsNumWidths]uint32, rng uint64) {
	for i := 0; i < BitsNumWidths; i++ {
		values[i] = uint32(rng>>i) & bitsMask(bitsWidths[i])
	}
}

// the single untimed WriteBits call site (§3.2)
func bitsWritePass(buffer []byte, values *[BitsNumWidths]uint32) int64 {
	w := serialize.NewBitWriter(buffer)
	for w.BitsAvailable() >= 256 {
		for i := 0; i < BitsNumWidths; i++ {
			w.WriteBits(values[i], bitsWidths[i])
		}
	}
	w.FlushBits()
	return w.BytesWritten()
}

// the single untimed ReadBits call site (§3.2): the buffer must read back
// exactly the values written — the bits family's refusal gate
func bitsReadVerify(buffer []byte, values *[BitsNumWidths]uint32) bool {
	r := serialize.NewBitReader(buffer)
	for r.BitsRemaining() >= 256 {
		for i := 0; i < BitsNumWidths; i++ {
			if r.ReadBits(bitsWidths[i]) != values[i] {
				return false
			}
		}
	}
	return true
}

//go:noinline
func bitpackerWriteLoop(passes int64, bytesPerPass int64, rng *uint64, values *[BitsNumWidths]uint32) bool {
	w := serialize.NewBitWriter(gBitsBuffer[:])
	for pass := int64(0); pass < passes; pass++ {
		*rng = benchRng(*rng)
		varyBitsValues(values, *rng)
		w.Reset(gBitsBuffer[:])
		for w.BitsAvailable() >= 256 {
			for i := 0; i < BitsNumWidths; i++ {
				w.WriteBits(values[i], bitsWidths[i])
			}
		}
		w.FlushBits()
		if w.BytesWritten() != bytesPerPass {
			return false // the bytes_per_op assertion (§2.7)
		}
		gSink = gSink + uint64(w.BytesWritten())
	}
	return true
}

//go:noinline
func bitpackerReadLoop(passes int64) bool {
	r := serialize.NewBitReader(gBitsVariants[0][:])
	for pass := int64(0); pass < passes; pass++ {
		r.Reset(gBitsVariants[pass&(NumVariants-1)][:])
		sum := uint64(0)
		for r.BitsRemaining() >= 256 {
			for i := 0; i < BitsNumWidths; i++ {
				sum += uint64(r.ReadBits(bitsWidths[i]))
			}
		}
		gSink = gSink + sum
	}
	return true
}

func benchBitpacker(basePasses int64) {
	passes := basePasses
	var values [BitsNumWidths]uint32

	rng := uint64(1)
	bytesPerPass := int64(-1)
	for k := 0; k < NumVariants; k++ {
		rng = benchRng(rng)
		varyBitsValues(&values, rng)
		wrote := bitsWritePass(gBitsVariants[k][:], &values)
		if bytesPerPass < 0 {
			bytesPerPass = wrote
		}
		if wrote != bytesPerPass {
			fail("bitpacker", "variation changed bytes/pass — widths are the structure and must stay fixed")
			return
		}
		if !bitsReadVerify(gBitsVariants[k][:], &values) {
			fail("bitpacker", "read-back disagrees with written values — refusing to bench")
			return
		}
	}

	writeRates := make([]float64, gNumRuns)
	readRates := make([]float64, gNumRuns)

	for run := -1; run < gNumRuns; run++ {
		start := time.Now()
		if !bitpackerWriteLoop(passes, bytesPerPass, &rng, &values) {
			fail("bitpacker", "bytes/pass changed in the timed loop (§2.7 assertion)")
			return
		}
		elapsed := time.Since(start).Seconds()
		if run >= 0 {
			writeRates[run] = float64(passes) / elapsed
		}
	}

	for run := -1; run < gNumRuns; run++ {
		start := time.Now()
		if !bitpackerReadLoop(passes) {
			fail("bitpacker", "read loop failed")
			return
		}
		elapsed := time.Since(start).Seconds()
		if run >= 0 {
			readRates[run] = float64(passes) / elapsed
		}
	}

	report("bitpacker", "write", passes, bytesPerPass, stats(writeRates), "bits")
	report("bitpacker", "read", passes, bytesPerPass, stats(readRates), "bits")
}

File: bench/rust/src/bits.rs
// schema bench — family bits for the Rust runner.
//
// Family bits (BENCH-STANDARD.md §1.4): the raw BitWriter/BitReader with the
// 16-width table (227 bits/group) over a 65536-byte buffer. Values vary per
// pass through the LCG (widths are the structure and stay fixed; bytes/pass
// asserted constant); reads rotate 64 pre-written variant buffers, each
// verified to read back exactly what was written before any number is
// produced. No `pub` beyond the driver entry point per §3.1; the timed loops
// are #[inline(never)] so the §4.1 verdict has a loop body to count.

use std::hint::black_box;
use std::time::Instant;

use serialize::{BitReader, BitWriter};

use crate::{Ctx, MAX_NUM_RUNS, NUM_VARIANTS, bench_rng, run_stats};

const BITS_NUM_WIDTHS: usize = 16;
const BITS_WIDTHS: [u32; BITS_NUM_WIDTHS] =
    [1, 32, 7, 13, 3, 25, 8, 19, 4, 28, 11, 16, 2, 30, 6, 22]; // 227 bits/group
const BITS_BUFFER_SIZE: usize = 65536;

fn bits_mask(width: u32) -> u32 {
    if width == 32 {
        0xFFFFFFFF
    } else {
        (1u32 << width) - 1
    }
}

// the per-pass value variation: one LCG step per pass, values from its bits
fn vary_bits_values(values: &mut [u32; BITS_NUM_WIDTHS], rng: u64) {
    for i in 0..BITS_NUM_WIDTHS {
        values[i] = (rng >> i) as u32 & bits_mask(BITS_WIDTHS[i]);
    }
}

// the single untimed write_bits call site (§3.2)
fn bits_write_pass(buffer: &mut [u8], values: &[u32; BITS_NUM_WIDTHS]) -> i64 {
    let mut w = BitWriter::new(buffer);
    while w.bits_available() >= 256 {
        for i in 0..BITS_NUM_WIDTHS {
            w.write_bits(values[i], BITS_WIDTHS[i]);
        }
    }
    w.flush_bits();
    w.bytes_written() as i64
}

// the single untimed read_bits call site (§3.2): the buffer must read back
// exactly the values written — the bits family's refusal gate
fn bits_read_verify(buffer: &[u8], values: &[u32; BITS_NUM_WIDTHS]) -> bool {
    let mut r = BitReader::new(buffer, BITS_BUFFER_SIZE);
    while r.bits_remaining() >= 256 {
        for i in 0..BITS_NUM_WIDTHS {
            if r.read_bits(BITS_WIDTHS[i]) != values[i] {
                return false;
            }
        }
    }
    true
}

#[inline(never)]
fn bitpacker_write_loop(
    ctx: &Ctx,
    passes: i64,
    bytes_per_pass: i64,
    rng: &mut u64,
    values: &mut [u32; BITS_NUM_WIDTHS],
    buffer: &mut [u8],
) -> bool {
    for _ in 0..passes {
        *rng = bench_rng(*rng);
        vary_bits_values(values, *rng);
        let wrote;
        {
            let mut w = BitWriter::new(buffer);
            while w.bits_available() >= 256 {
                for i in 0..BITS_NUM_WIDTHS {
                    w.write_bits(values[i], BITS_WIDTHS[i]);
                }
            }
            w.flush_bits();
            wrote = w.bytes_written() as i64;
        }
        if wrote != bytes_per_pass {
            return false; // the bytes_per_op assertion (§2.7)
        }
        black_box(&*buffer);
        ctx.sink.set(ctx.sink.get().wrapping_add(wrote as u64));
    }
    true
}

#[inline(never)]
fn bitpacker_read_loop(ctx: &Ctx, passes: i64, variants: &[Vec<u8>]) -> bool {
    for pass in 0..passes {
        let mut r = BitReader::new(
            &variants[(pass as usize) & (NUM_VARIANTS - 1)],
            BITS_BUFFER_SIZE,
        );
        let mut sum: u64 = 0;
        while r.bits_remaining() >= 256 {
            for i in 0..BITS_NUM_WIDTHS {
                sum += r.read_bits(BITS_WIDTHS[i]) as u64;
            }
        }
        ctx.sink.set(ctx.sink.get().wrapping_add(sum));
    }
    true
}

pub fn bench_bitpacker(ctx: &Ctx, passes: i64) {
    let mut values = [0u32; BITS_NUM_WIDTHS];
    let mut buffer = vec![0u8; BITS_BUFFER_SIZE];
    let mut variants: Vec<Vec<u8>> = (0..NUM_VARIANTS)
        .map(|_| vec![0u8; BITS_BUFFER_SIZE])
        .collect();

    let mut rng: u64 = 1;
    let mut bytes_per_pass: i64 = -1;
    for k in 0..NUM_VARIANTS {
        rng = bench_rng(rng);
        vary_bits_values(&mut values, rng);
        let wrote = bits_write_pass(&mut variants[k], &values);
        if bytes_per_pass < 0 {
            bytes_per_pass = wrote;
        }
        if wrote != bytes_per_pass {
            ctx.fail(
                "bitpacker",
                "variation changed bytes/pass — widths are the structure and must stay fixed",
            );
            return;
        }
        if !bits_read_verify(&variants[k], &values) {
            ctx.fail(
                "bitpacker",
                "read-back disagrees with written values — refusing to bench",
            );
            return;
        }
    }

    let mut write_rates = [0.0f64; MAX_NUM_RUNS];
    let mut read_rates = [0.0f64; MAX_NUM_RUNS];

    for run in 0..(ctx.num_runs + 1) {
        let start = Instant::now();
        if !bitpacker_write_loop(
            ctx,
            passes,
            bytes_per_pass,
            &mut rng,
            &mut values,
            &mut buffer,
        ) {
            ctx.fail(
                "bitpacker",
                "bytes/pass changed in the timed loop (§2.7 assertion)",
            );
            return;
        }
        let time = start.elapsed().as_secs_f64();
        if run >= 1 {
            write_rates[run - 1] = passes as f64 / time;
        }
    }

    for run in 0..(ctx.num_runs + 1) {
        let start = Instant::now();
        if !bitpacker_read_loop(ctx, passes, &variants) {
            ctx.fail("bitpacker", "read loop failed");
            return;
        }
        let time = start.elapsed().as_secs_f64();
        if run >= 1 {
            read_rates[run - 1] = passes as f64 / time;
        }
    }

    ctx.report(
        "bitpacker",
        "write",
        passes,
        bytes_per_pass,
        &run_stats(&mut write_rates[..ctx.num_runs]),
        "bits",
    );
    ctx.report(
        "bitpacker",
        "read",
        passes,
        bytes_per_pass,
        &run_stats(&mut read_rates[..ctx.num_runs]),
        "bits",
    );
}
